internal/devconfig/packages.go

```diff
@@ -0,0 +1,226 @@
+	// specification specifies that maps are unordered, we do rely on the order
+	// for certain functionality.
+	orderedMap := orderedmap.New[string, Package]()
+	err := json.Unmarshal(data, &orderedMap)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	// Convert the ordered map to a list of packages, and set the name field
+	// from the map's key
+	packagesList := []Package{}
+	for pair := orderedMap.Oldest(); pair != nil; pair = pair.Next() {
+		pkg := pair.Value
+		pkg.name = pair.Key
+		packagesList = append(packagesList, pkg)
+	}
+	pkgs.Collection = packagesList
+	pkgs.jsonKind = jsonMap
+	return nil
+}
+
+func (pkgs *Packages) MarshalJSON() ([]byte, error) {
+	if pkgs.jsonKind == jsonList {
+		packagesList := []string{}
+		for _, p := range pkgs.Collection {
+
+			// Version may be empty for unversioned packages
+			packageToWrite := p.name
+			if p.Version != "" {
+				packageToWrite += "@" + p.Version
+			}
+			packagesList = append(packagesList, packageToWrite)
+		}
+		return json.Marshal(packagesList)
+	}
+
+	orderedMap := orderedmap.New[string, Package]()
+	for _, p := range pkgs.Collection {
+		orderedMap.Set(p.name, p)
+	}
+	return json.Marshal(orderedMap)
+}
+
+type packageKind int
+
+const (
+	versionOnly packageKind = iota
+	regular     packageKind = iota
+)
+
+type Package struct {
+	kind packageKind
+	name string
+	// deliberately not adding omitempty
+	Version string `json:"version"`
+
+	// TODO: add other fields like platforms
+}
+
+func NewVersionOnlyPackage(name, version string) Package {
+	return Package{
+		kind:    versionOnly,
+		name:    name,
+		Version: version,
+	}
+}
+
+func NewPackage(name string, values map[string]any) Package {
+	version, ok := values["version"]
+	if !ok {
+		// For legacy packages, the version may not be specified. We leave it blank
+		// here, and code that consumes the Config is expected to handle this case
+		// (e.g. by defaulting to @latest).
+		version = ""
+	}
+
+	return Package{
+		kind:    regular,
+		name:    name,
+		Version: version.(string),
+	}
+}
+
+func (p *Package) UnmarshalJSON(data []byte) error {
+	// First, attempt to unmarshal as a version-only string
+	var version string
+	if err := json.Unmarshal(data, &version); err == nil {
+		p.kind = versionOnly
+		p.Version = version
+		return nil
+	}
+
+	// Second, attempt to unmarshal as a Package struct
+	type Alias Package // Use an alias-type to avoid infinite recursion
+	alias := &Alias{}
+	if err := json.Unmarshal(data, alias); err != nil {
+		return errors.WithStack(err)
+	}
+
+	// more robust way to copy all fields from alias?
+	p.kind = regular
+	p.Version = alias.Version
+	return nil
+}
+
+func (p Package) MarshalJSON() ([]byte, error) {
+	if p.kind == versionOnly {
+		return json.Marshal(p.Version)
+	}
+
+	// If we have a regular package, we want to marshal the entire struct:
+	type Alias Package // Use an alias-type to avoid infinite recursion
```

Comment:
This is cool! I've gotten into infinite recursion issue when implementing custom marshaling and I don't remember the solution being this clean.

nit, make lower case